src/modules/ekf2/ekf2_main.cpp

```diff
@@ -1191,38 +1165,43 @@ void Ekf2::run()
 					}
 				}
 
-				// Check and save the last valid calibration when we are disarmed
-				if ((vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_STANDBY)
-				    && (status.filter_fault_flags == 0)
+				// Check and update the last valid magnetometer calibration
+				if ((status.filter_fault_flags == 0)
 				    && (sensor_selection.mag_device_id == _mag_bias_id.get())) {
 					control::BlockParamFloat *mag_biases[] = { &_mag_bias_x, &_mag_bias_y, &_mag_bias_z };
 
 					for (uint8_t axis_index = 0; axis_index <= 2; axis_index++) {
+						// update bias estimates
 						if (_valid_cal_available[axis_index]) {
 							// calculate weighting using ratio of variances and update stored bias values
 							float weighting = _mag_bias_saved_variance.get() / (_mag_bias_saved_variance.get() +
 									  _last_valid_variance[axis_index]);
 							weighting = math::constrain(weighting, 0.0f, _mag_bias_alpha.get());
 							float mag_bias_saved = mag_biases[axis_index]->get();
 							_last_valid_mag_cal[axis_index] = weighting * _last_valid_mag_cal[axis_index] + mag_bias_saved;
-							mag_biases[axis_index]->set(_last_valid_mag_cal[axis_index]);
-							mag_biases[axis_index]->commit_no_notification();
+
+							// Only save parameters when disarmed
+							if (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_STANDBY) {
+								mag_biases[axis_index]->set(_last_valid_mag_cal[axis_index]);
+								mag_biases[axis_index]->commit_no_notification();
+
+								// reset to prevent data being saved too frequently
+								_total_cal_time_us = 0;
+							}
 
 							_valid_cal_available[axis_index] = false;
 						}
 					}
-
-					// reset to prevent data being saved too frequently
-					_total_cal_time_us = 0;
 				}
 
 				// Publish wind estimate
 				struct wind_estimate_s wind_estimate = {};
 				wind_estimate.timestamp = now;
 				wind_estimate.windspeed_north = status.states[22];
```

Comment:
@priseborough I think it will be better if you added getters here instead of accessing the states directly like this.